dht dynamic mode switching.

raulk opened the pull request:
We now listen to local routability events to dynamically switch our DHT mode. 

Private/unknown routability transition us to client mode; public routability puts us in server mode.

Fixes #216.

whyrusleeping commented on dht.go:
```diff
@@ -119,14 +120,20 @@ func New(ctx context.Context, h host.Host, options ...opts.Option) (*IpfsDHT, er
 
-		if dht.subscriptions.evtPeerIdentification != nil {
-			_ = dht.subscriptions.evtPeerIdentification.Close()
+		for _, sub := range []event.Subscription{
```
lol, creative way to avoid an extra `!= nil` check without writing less code :P

raulk replied:
I could remove the line breaks between slice elements to reduce line count :P

whyrusleeping commented on dht.go:
```diff
@@ -498,6 +522,33 @@ func (dht *IpfsDHT) SetMode(m DHTMode) error {
+			case ev := <-dht.subscriptions.evtLocalRoutability.Out():
+				var err error
+				switch ev.(type) {
+				case event.EvtLocalRoutabilityPrivate, event.EvtLocalRoutabilityUnknown:
```
I'm a little nervous about flapping here. Could we add a check that this hasnt happened too recently, and throw an error log if it does?

Stebalien replied:
We could also debounce this:

1. Set a local "routable" variable each time.
2. Reset a timeout each time we receive an event.
3. Only change modes when we haven't received an event for N seconds.

raulk replied:
The debouncing sounds sensible. What do we want the value of N to be? 1 minute?

raulk replied:
Implemented.

raulk commented:
@Stebalien @whyrusleeping debouncing implemented. Can I get a quick sanity check?

Stebalien commented on dht.go:
```diff
@@ -187,16 +198,33 @@ func makeDHT(ctx context.Context, h host.Host, dstore ds.Batching, protocols []p
 
+	dht.setupEventSubscribers()
+
+	dht.ctx = dht.newContextWithLocalTags(ctx)
```
Is this the right order? Subscribe then setup the context?

raulk replied:
event subscribers don't depend on the context, but logically it should be reversed, correct

Stebalien commented on dht.go:
```diff
@@ -498,6 +526,46 @@ func (dht *IpfsDHT) SetMode(m DHTMode) error {
+					target = ModeServer
+				}
+				if debounce == nil {
+					debounce = time.After(DynamicModeSwitchDebouncePeriod)
```
This is a throttle. That works but a debounce would probably be better. To do that, we'd:

1. Have a `time.Timer`.
2. Call `myTimer.Reset(deadline)` every time we receive an event.

whyrusleeping replied:
And if we have a debounce set, and we switch back to the mode we were previously in, just cancel the operation and clear out the timer.

raulk replied:
Actually debouncing now ;-) https://github.com/libp2p/go-libp2p-kad-dht/pull/367/commits/87f6b0c908de352e67034476588d2bc67171aa60

raulk commented:
Do not review yet, I'm cleaning this up.

Stebalien commented on dht.go:
```diff
@@ -495,6 +526,54 @@ func (dht *IpfsDHT) SetMode(m DHTMode) error {
+		}
+		select {
+		case <-debouncer.C:
+		default:
```
Note: this is technically kind of racy because Stop doesn't wait for the timer to stop, but it shouldn't really be an issue.

raulk replied:
timer.Stop() is a huge headache. I never got it to work as described in godocs, so I went this route 🙃

Stebalien commented:
@raulk I made `dynamicModeSwitching` just use `goprocess.Go` directly. Feel free to blow that away if you're refactoring (we can just redo it).

raulk commented:
> @raulk I made dynamicModeSwitching just use goprocess.Go directly. Feel free to blow that away if you're refactoring (we can just redo it).

@Stebalien 	exactly the refactoring I was doing. I went a little further and fixed `subscriberNotifee` too to use the same approach. I also moved the subscriptions to local scope.

Stebalien commented on subscriber_notifee.go:
```diff
@@ -16,23 +17,27 @@ func (nn *subscriberNotifee) DHT() *IpfsDHT {
 	dht := nn.DHT()
 
-	proc := goprocess.Go(nn.subscribe)
 	dht.host.Network().Notify(nn)
```
Actually, this applies to the other subscribers as well.

aarshkshah1992 replied:
@Stebalien  Please can you elaborate on this ?

Stebalien replied:
I can't remember... I think my concern was that we could listen for new connections after the network has started and we already have connections, but that's already the case.

aschmahmann commented:
@aarshkshah1992  it looks like you're taking another stab at this PR. I don't think that the debouncer stuff necessarily makes a ton of sense, it should (and to some extent already does) exist in autonat. Adding a separate one for the DHT seems a little strange.

I ported some of these changes into my DHT branch (https://github.com/libp2p/go-libp2p-kad-dht/pull/436/commits/35d3e4a5d43eaaf9f4cab77b6570be35721d66d9).  I wasn't sure what was going on with the peer event changes removal changes in this PR so I left it alone.

aarshkshah1992 commented:
Code from this PR has been absorbed in libp2p/go-libp2p-kad-dht#436 & libp2p/go-libp2p-kad-dht#448 combined.  This PR can now be closed.
